### PR TITLE
Upgrade to CosmosDb SDK v3

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description
Removes the v2 SDK and upgrades all apis to use v3.
Going forward this has functionality such as Batches/Transactions/Bulk apis that we can look into utilizing.

Also merges the two *Fhir.CosmosDb and *.CosmosDb projects back together

## Related issues
Addresses [AB#74576](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/74576).

## Testing
Converts all existing tests to use new API

### CODE REVIEW

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosClientInitializer.cs [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+                // By default, the Json.NET serializer uses 'F' instead of 'f' for fractions.
+                // 'F' will omit the trailing digits if they are 0. You might end up getting something like '2018-02-07T20:04:49.97114+00:00'
+                // where the fraction is actually 971140. Because the ordering is done as string,
+                // if the values don't always have complete 7 digits, the comparison might not work properly.
+                serializerSettings.Converters.Add(new IsoDateTimeConverter { DateTimeFormat = "o" });
+
+                return JsonSerializer.Create(serializerSettings);
+            }
+
+            public override T FromStream<T>(Stream stream)
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosClientInitializer.cs:174]: Open to perf improvements / suggestions for the To/From stream functions

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosClientInitializer.cs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            return builder.Build();
+        }
+
+        public Container CreateFhirContainer(CosmosClient client, string databaseId, string collectionId, int? continuationTokenSizeLimitInKb)
+        {
+            return client.GetContainer(databaseId, collectionId);
+        }
+
+        /// <inheritdoc />
+        public async Task OpenDocumentClient(CosmosClient client, CosmosDataStoreConfiguration configuration, CosmosCollectionConfiguration cosmosCollectionConfiguration)
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosClientInitializer.cs]: Should we rename this?

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/Registry/CosmosDbStatusRegistryInitializer.cs [hunk 3]
```diff
@@ -30,28 +30,33 @@ public class CosmosDbStatusRegistryInitializer : IFhirCollectionUpdater
...
 
             if (!results.Any())
             {
                 var statuses = await _filebasedRegistry.GetSearchParameterStatuses();
 
-                foreach (SearchParameterStatusWrapper status in statuses.Select(x => x.ToSearchParameterStatusWrapper()))
+                foreach (var batch in statuses.TakeBatch(100))
                 {
-                    await client.UpsertDocumentAsync(relativeCollectionUri, status);
+                    TransactionalBatch transaction = collection.CreateTransactionalBatch(new PartitionKey(SearchParameterStatusWrapper.SearchParameterStatusPartitionKey));
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/Registry/CosmosDbStatusRegistryInitializer.cs:52]: Reuse the partitionkey instance?

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/Registry/CosmosDbStatusRegistryInitializer.cs [hunk 4]
```diff
@@ -30,28 +30,33 @@ public class CosmosDbStatusRegistryInitializer : IFhirCollectionUpdater
...
-            EnsureArg.IsNotNull(relativeCollectionUri, nameof(relativeCollectionUri));
 
             // Detect if registry has been initialized
             var query = _queryFactory.Create<dynamic>(
-                client,
+                collection,
                 new CosmosQueryContext(
-                    relativeCollectionUri,
-                    new SqlQuerySpec($"SELECT TOP 1 * FROM c where c.{KnownDocumentProperties.PartitionKey} = '{SearchParameterStatusWrapper.SearchParameterStatusPartitionKey}'")));
+                    new QueryDefinition($"SELECT TOP 1 * FROM c where c.{KnownDocumentProperties.PartitionKey} = '{SearchParameterStatusWrapper.SearchParameterStatusPartitionKey}'"),
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/Registry/CosmosDbStatusRegistryInitializer.cs:41]: Do we need to specoty the partition in the query text as well as the QueryRequestOptions?

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/StoredProcedures/IStoredProcedure.cs [hunk 5]
```diff
@@ -3,11 +3,14 @@
...
+using Microsoft.Azure.Cosmos.Scripts;
 
 namespace Microsoft.Health.Fhir.CosmosDb.Features.Storage.StoredProcedures
 {
-    public interface IFhirStoredProcedure : IStoredProcedure
+    public interface IStoredProcedure
     {
+        string FullName { get; }
+
+        StoredProcedureProperties AsStoredProcedure();
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/StoredProcedures/IStoredProcedure.cs:14]: Maybe `AsStoredProcedureProperties()` or `GetStoredProcedureProperties()`

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosDbDistributedLock.cs [hunk 6]
```diff
@@ -8,14 +8,13 @@
...
-using Microsoft.Azure.Documents;
-using Microsoft.Azure.Documents.Client;
+using Microsoft.Azure.Cosmos;
 using Microsoft.Extensions.Logging;
 using Microsoft.Health.Abstractions.Exceptions;
 using Microsoft.Health.Extensions.DependencyInjection;
 using Newtonsoft.Json;
 
-namespace Microsoft.Health.CosmosDb.Features.Storage
+namespace Microsoft.Health.Fhir.CosmosDb.Features.Storage
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosDbDistributedLock.cs:17]: This class would actually benefit from being `IAsyncDisposable`

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosResponseProcessor.cs [hunk 7]
```diff
@@ -42,61 +41,44 @@ public CosmosResponseProcessor(IFhirRequestContextAccessor fhirRequestContextAcc
         /// Adds request charge to the response headers and throws a <see cref="RequestRateExceededException"/>
         /// if the status code is 429.
         /// </summary>
-        /// <param name="ex">The exception</param>
-        public async Task ProcessException(Exception ex)
+        /// <param name="response">The response that has errored</param>
+        public Task ProcessException(ResponseMessage response)
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosResponseProcessor.cs:45]: Maybe ProcessErrorResponse?

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosClientInitializer.cs [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+                using var writer = new NonClosingStreamWriter(stream);
+                using var jsonWriter = new JsonTextWriter(writer);
+                _serializer.Serialize(jsonWriter, input);
+                jsonWriter.Flush();
+                stream.Position = 0;
+                return stream;
+            }
+        }
+
+        private class NonClosingStreamWriter : StreamWriter
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosClientInitializer.cs:193]: You can instead specify that a StreamWriter leave the stream open by a constructor parameter.

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosReponseHandler.cs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+using Microsoft.Azure.Cosmos;
+using Microsoft.Health.Extensions.DependencyInjection;
+using Microsoft.Health.Fhir.Core.Features.Context;
+using Microsoft.Health.Fhir.Core.Features.Persistence;
+using Microsoft.Health.Fhir.CosmosDb.Configs;
+using Microsoft.Health.Fhir.CosmosDb.Features.Queries;
+
+namespace Microsoft.Health.Fhir.CosmosDb.Features.Storage
+{
+    public class FhirCosmosReponseHandler : RequestHandler
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosReponseHandler.cs:20]: typo

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosReponseHandler.cs [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+                options.Headers[_consistencyLevelHeaderName] = consistencyLevel?.ToString();
+            }
+
+            if (!string.IsNullOrEmpty(sessionToken))
+            {
+                options.Headers[_sessionTokenHeaderName] = sessionToken;
+            }
+        }
+
+        private (ConsistencyLevel? consistencyLevel, string sessionToken) GetConsistencyHeaders()
```
Comment (johnstairs) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/FhirCosmosReponseHandler.cs:85]: This is so much better than having to resort to codegen!

File: src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/Versioning/FhirCollectionSettingsUpdater.cs [hunk 11]
```diff
@@ -45,61 +39,55 @@ public sealed class FhirCollectionSettingsUpdater : IFhirCollectionUpdater
...
-                    },
-                    ExcludedPaths = new Collection<ExcludedPath>
-                    {
-                        new ExcludedPath
-                        {
-                            Path = $"/{KnownResourceWrapperProperties.RawResource}/*",
-                        },
-                    },
-                };
+                    Path = $"/{KnownResourceWrapperProperties.RawResource}/*",
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/Versioning/FhirCollectionSettingsUpdater.cs:66]: Possibly also add the default `"path": "/\"_etag\"/?"` here too

**Pull request merged**